Automattic/jetpack
# Search: fix translations for the new search customizer and lazy loaded payload

Fixes #21349 and #20520

#### Changes proposed in this Pull Request:

##### Background

> Trying it locally, I see this in my browser console:
> Uncaught (in promise) Error: Loading CSS chunk main-payload failed.
(https://REDACTED/wp-content/plugins/jetpack/_inc/build/instant-search/jp-search.chunk-main-payload.min.rtl.css?ver=f33ebf8feb95ae7fc64a)
> [...] @automattic/webpack-rtl-plugin expects the CSS asset's name to end in .css, while since #20926 it's ending in something like .css?ver=f33ebf8feb95ae7fc64a. So it never produces the .rtl.css file, which makes Instant Search fail to load for RTL languages. (@anomiex)

In #20926, we moved the `contentHash` from file name to [a GET parameter](https://github.com/Automattic/jetpack/pull/20926/files#diff-334181bb7a4c7626231d3a9e6d5aaa8d3a771d116848e2c26b1c9b596513e3d4R30) which breaks `rtl` languages like Arabic (#21349).

>At least wp i18n make-pot skips .min.js files by default, which we have to use to work around wpcom's broken minifier. [...] I don't know whether the official translation mechanism does the same. (@anomiex)

And also, when extracting string for translation, asset names ending with `.min.js` are all ignored (p1632856853438200-slack-jetpack-crew). Customburg (the search customizer) and lazy-loaded assets are all with `.min.js`, so strings in them are not translated at all.

##### Changes

The PR is basically a follow-up for #20926, which only proposed a solution to load translations but didn't make sure the translations would be generated. The PR proposes a Webpack plugin, which emits readable JS assets along with minimized assets. The plugin also removes the hash in the un-minimized asset filename, the reasons being that Jetpack the plugin could enqueue translations for a fixed filename rather than one changing all the time, which is essential for WPCOM as we deploy frequently with the hash changing all the time. The downside of doing this is Jetpack release will become about 800KB larger. The original plugin could be found [here](https://www.npmjs.com/package/@wordpress/readable-js-assets-webpack-plugin). Thanks for pointing the way @anomiex. 

Also, this PR reverted the change of moving the `contentHash` from file name to a GET parameter.

In the future, we could possibly automate the process by generating a list of assets for the PHP to automatically inline translations. Another item to investigate: if the RTL CSS build failed, why were no errors thrown?

#### Jetpack product discussion
pcNPJE-gU-p2

#### Does this pull request change what data or activity we track or use?
no

#### Testing instructions:
- Run `pnpm build-production-search` 
- Check files in `_inc/build/instant-search`
- Ensure for `.min.js` files, there are un-minimized files for each minimized file.
- Ensure for `jp-search.chunk-main-payload.[hash].min.js`, the un-minimized version is `jp-search.chunk-main-payload.js`
- Ensure Search and Customburg work fine


- Run `cd path/to/jetpack/the/plugin && wp i18n make-pot ./ jetpack.pot`
- Ensure strings from `jp-search.chunk-main-payload.js`, `jp-search-main.bundle.js` and `jp-search-configure-main.js` are correctly extracted. Example:
```
#: _inc/build/instant-search/jp-search-configure-main.js:4542
msgid "Options"
msgstr ""
```
( You might want to [install wp for your mac](https://wpbeaches.com/installing-wordpress-wp-cli-macos/) as the docker instance doesn't have enough memory. )


- Download [the mocked French language file](https://github.com/Automattic/jetpack/files/7262897/jetpack-fr_FR-0e17ef273599d7b1004fc0c6aacf4f5e.json.zip)
- Unzip the file to `tools/docker/wordpress/wp-content/languages/plugins` and then switch your WP language to `French`
- Open the search overlay, and ensure you see `In French: Showing popular results` like the screenshot below.

![image](https://user-images.githubusercontent.com/1425433/135539955-8077789c-01e0-4200-bd4f-84d8bb0ababa.png)
